python/ray/data/_internal/execution/operators/map_operator.py

```diff
@@ -49,75 +87,404 @@ def __init__(
+                transform_fn,
+                input_op,
+                name=name,
+                min_rows_per_bundle=min_rows_per_bundle,
+                ray_remote_args=ray_remote_args,
+                pool_size=pool_size,
+            )
+        else:
+            raise ValueError(f"Unsupported execution strategy {compute_strategy}")
+
+    def start(self, options: "ExecutionOptions"):
+        # Create output queue with desired ordering semantics.
+        if options.preserve_order:
+            self._output_queue = _OrderedOutputQueue()
+        else:
+            self._output_queue = _UnorderedOutputQueue()
+        if options.locality_with_output:
+            # Try to schedule tasks locally.
+            self._ray_remote_args[
+                "scheduling_strategy"
+            ] = NodeAffinitySchedulingStrategy(
+                ray.get_runtime_context().get_node_id(),
+                soft=True,
+            )
+        super().start(options)
 
-    def add_input(self, refs: RefBundle, input_index: int) -> None:
+    def add_input(self, refs: RefBundle, input_index: int):
         assert input_index == 0, input_index
-        self._execution_state.add_input(refs)
+        # Add RefBundle to the bundler.
+        self._block_ref_bundler.add_bundle(refs)
+        if self._block_ref_bundler.has_bundle():
+            # If the bundler has a full bundle, add it to the operator's task submission
+            # queue.
+            bundle = self._block_ref_bundler.get_next_bundle()
+            self._add_bundled_input(bundle)
+
+    @abstractmethod
+    def _add_bundled_input(self, refs: RefBundle):
+        """Add a pre-bundled upstream output to this operator.
+
+        Unlike the add_input() arg, this RefBundle has already been further bundled by
+        _block_ref_bundler up to the target size, meaning that this bundle is ready for
+        task submission.
+
+        This must be implemented by subclasses.
+
+        Args:
+            refs: The fully-bundled ref bundle that should be added as input.
+        """
+        raise NotImplementedError
+
+    def _handle_task_submitted(self, task: "_TaskState"):
+        """Handle a newly submitted task, notifying the output queue and updating
+        object store metrics.
+
+        This should be called by subclasses right after a task is submitted.
+
+        Args:
+            task: The task state for the newly submitted task.
+        """
+        # Notify output queue that this task is pending.
+        self._output_queue.notify_pending_task(task)
+        # Update object store metrics.
+        self._metrics.cur += task.inputs.size_bytes()
+        if self._metrics.cur > self._metrics.peak:
+            self._metrics.peak = self._metrics.cur
+
+    @abstractmethod
+    def notify_work_completed(
+        self, ref: Union[ObjectRef[ObjectRefGenerator], ray.ObjectRef]
+    ):
+        """Indicates that a task is done executing OR that a worker is done starting.
```

Comment:
Why "task is done" is not already handled by `_handle_task_done`? It's unclear how `notify_work_completed` is different. Also it looks handling dual responsibilities (worker done starting part).

Reply:
`notify_work_completed` is the `PhysicalOperator` API for "this future that you wanted the executor is done"; this includes both task futures AND worker is ready futures, and is an intentional part of the `Executor` <--> `PhysicalOperator` API and contract (notice that it's `notify_work_completed()` and `get_work_refs()`, not `notify_task_completed()` and `get_task_refs()`).

The current contract is: a `PhysicalOperator` provides a list of futures that it wants to executor's control loop to wait on (`op.get_work_refs()`), expecting the executor to tell the operator when each is done (`op.notify_work_completed()`). The executor doesn't care what these futures are, and the `PhysicalOperator` can have those futures represent any work it needs to carry out, as long as it can be waited on with `ray.wait()`: this can be an executing task, actor starting up, a placement group being scheduled, or even waiting for something external to the Ray cluster to be ready. The `ActorPoolMapOperator` exposes both running task futures and pending actor futures in `op.get_work_refs()`, and delineates between the two when a ref is received in `op.notify_work_completed()`.

We could have a separate `op.get_worker_refs()` and `op.notify_worker_started()` APIs, but that's exposing unnecessary detail to the `Executor`, who doesn't need to know those kind of details about the futures that its waiting on.

cc @ericl for another perspective on this

Reply:
Maybe we should rename it to on_ray_event or something in another PR.

Reply:
Thanks for clarifying. Can we add in comment about what is "work", which you mentioned already it's all things that can be `ray.wait()` on?.

Reply:
@jianoaix We have docstring comments in `MapOperator.notify_work_completed` and `ActorPoolMapOperator.notify_work_completed` that "work completed" can mean that either a task is done executing or a worker is done starting, are you referring to adding more generic docs to `PhysicalOperator.notify_work_completed()` and `PhysicalOperator.get_work_refs()`?